### ISSUE TITLE
polygon tool issue state 791

### ISSUE THREAD

**Issue by befriedman** [ActiveBrainAtlas2/activebrainatlasadmin#228]
<img width="697" alt="image" src="https://github.com/ActiveBrainAtlas2/activebrainatlasadmin/assets/59492402/37e33b63-74a9-4647-807d-120304b5cbca">

I cannot batch change polygon color and it is set to be the same as the cell annotation color
These need to be distinct
Can this be fixed as for some reason it was misassigned when created

Ideally the cells would be yellow
The polygons would be red

Also if I want to hide the polygons by striking through the polyon label 
<img width="697" alt="image" src="https://github.com/ActiveBrainAtlas2/activebrainatlasadmin/assets/59492402/ad8b0543-57ae-4d33-86f3-c6deb8813956">
they don't disappear

Thank you,
Beth

**Comment by befriedman** [comment]
<img width="528" alt="image" src="https://github.com/ActiveBrainAtlas2/activebrainatlasadmin/assets/59492402/0bacbe96-9f25-4b89-9a99-4002508a667d">
Corxn for image of strikethrough of polygon tab

**Comment by eddyod** [comment]
There are 3 layers with annotations. It seems there is some redundant data between the polygons layer and the 'annotations' layer. Try clicking either of them on and off.  As for the color, Since they have already been created red, you'll need to change them manually in the lower right panel.

**Comment by eddyod** [comment]
We need a way to batch update an annotation's layers color.

**Comment by drinehart1** [comment]
Another issue (for me) is the inability to audit the annotations.  If I am interpreting the situation correctly:

1) All information about the id state is available with:
SELECT created, updated, description, comments, username FROM neuroglancer_state a
INNER JOIN auth_user b ON b.id = a.FK_user_id
WHERE a.id=791

created	updated	description	comments	username
2023-02-13 14:51:22.784993	2023-02-13 14:52:58.000000	DK78 TG Polygons2 - Marissa	DK78 TG Polygons2 - Marissa	Marissa

-> Unfortunately this only tracks the create and update datestamps for setting up the layers, not annotations.  I find it hard to believe Marissa created all the annotations in <2 min.

2) All meta-data is stored in the field 'neuroglancer_state' (in table of same name).  For annotation layers that include polygons, this includes all the vertices and ordering by unique polygon, but the tracking of changes to the polygons e.g. moved vertices were supposed to be stored as user sessions in 'annotation_sessions' table.  

2a) This table has fields to store what changed: user, datestamps, coordinates, etc. but only populates when users click 'Save Annotations'.  Even then, I don't fully understand why we have 17.6% of annotations without any state id.  How did these rogue annotations get into table?

ref: SELECT (SELECT COUNT(*) FROM annotation_session WHERE annotation_type=1 AND FK_state_id IS NULL) / (SELECT COUNT(*) FROM annotation_session WHERE annotation_type=1 AND FK_state_id >0)

As for this specific id, there were not annotation sessions at all before I tested today:

ref: SELECT * FROM annotation_session WHERE FK_state_id=791

**Comment by drinehart1** [comment]
Before we can batch up annotations (to assign them custom colors or even standardize the colors to avoid confusion between sets), we should address the fact that Neuroglancer states convolute layers with annotation sets.  We may not need to modify the front-end UI (as this is already challenging and the goal is to stick close to open source version).  I propose the following:

We can leave most recent, authoritative annotation set in neuroglancer_state.neuroglancer_state (Not my preference as some data may be duplicated but easiest path forward)

For tracking changes there are options (visually I like https://timeline.knightlab.com/#examples to show end-users what changed) but we have to store the info first.

On load specific id from brainsharer page, we can use current source (in table).

Annotation save seems to work but when entire state is saved, there is no record of the changes.  Upon save state (in Django code):

parse annotations from url state (look for '"type": "annotation"' in json code) example:

data = json.loads(field_data)
for layer in data['layers']:
    if layer['type'] == 'annotation':
        print(layer['type'])

INSERT into annotations_session (meta-data for url state)
INSERT the actual changed data intopolygon_sequences WHERE FK_session_id={insert_id}

**Comment by drinehart1** [comment]
Once we have batched up the annotations, we can address the color issues:

a) update all annotations with same color (UI change) - split by polygon vs points
b) automatically suggest colors based on color difference calculations (https://python-colormath.readthedocs.io/en/latest/) to ensure user and visually distinguish different colors
c) allow user to use "standardized" color scheme (default colors for identified structures), which would not change across brains

**Comment by eddyod** [comment]
> Another issue (for me) is the inability to audit the annotations. If I am interpreting the situation correctly:
> 
> 1. All information about the id state is available with:
>    SELECT created, updated, description, comments, username FROM neuroglancer_state a
>    INNER JOIN auth_user b ON b.id = a.FK_user_id
>    WHERE a.id=791
> 
> created updated description comments username 2023-02-13 14:51:22.784993 2023-02-13 14:52:58.000000 DK78 TG Polygons2 - Marissa DK78 TG Polygons2 - Marissa Marissa
> 
> -> Unfortunately this only tracks the create and update datestamps for setting up the layers, not annotations. I find it hard to believe Marissa created all the annotations in <2 min.
> 
> 2. All meta-data is stored in the field 'neuroglancer_state' (in table of same name).  For annotation layers that include polygons, this includes all the vertices and ordering by unique polygon, but the tracking of changes to the polygons e.g. moved vertices were supposed to be stored as user sessions in 'annotation_sessions' table.
> 
> 2a) This table has fields to store what changed: user, datestamps, coordinates, etc. but only populates when users click 'Save Annotations'. Even then, I don't fully understand why we have 17.6% of annotations without any state id. How did these rogue annotations get into table?
> 
> ref: SELECT (SELECT COUNT(_) FROM annotation_session WHERE annotation_type=1 AND FK_state_id IS NULL) / (SELECT COUNT(_) FROM annotation_session WHERE annotation_type=1 AND FK_state_id >0)
> 
> As for this specific id, there were not annotation sessions at all before I tested today:
> 
> ref: SELECT * FROM annotation_session WHERE FK_state_id=791

The timestamps are the ones on the annotation session. When a user clicks 'Save annotations' The annotation session in that layer gets updated. It then appears as if those annotations all got updated at that same time. Same with the created column. The entire saving of annotations needs to be re-implemented as it is not robust. It is also being saved in 3 different places. I'm hoping Litao's updates with firestore will bring some improvements. We should address this when he has implemented his polygon tool.

**Comment by eddyod** [comment]
One way to update the colors is by following these steps. Assuming you have two layers with the same color:
1. Control click the layer you whose color you want to modify. 
2. Click the 'Save Annotations' button. 
3. Open up a new annotation layer (control + ).
4. Go to the drop down menu on the right and select the annotation type you just saved.
5. Find the cells (or polygons) you just saved 
6. Select a new different color
7. Import the cells under that color
8. Remove the old layer with the redundant color

**Comment by befriedman** [comment]
Hi Ed,

Does this require changing color cell by cell? If so then I think it is easier to just redo these annotations. If there is a way to alter the color of cells as a batch that would be preferable.

In any case the other possible fix would be to track back to a version where the color of the cells are displayed in the original yellow color-for example such a (correct) version was used for the last brain meeting when we did the live demo for neuroglancer.

...hope you found your phone- that is indeed a hassle

best,

Beth


________________________________

**Comment by eddyod** [comment]
When you import the cells, you choose the color ahead of time so when you import, they are all the new color.

**Comment by befriedman** [comment]
Hi Ed,

I think Marissa and I will need a refresher about importing the cells- it sounds like there is a fix

Can we arrange for a zoom call- maybe on Tuesday for your instruction?

Thanks so much,

Beth
________________________________

**Comment by befriedman** [comment]
Hi Beth,

Link id 791 is an incomplete version of the finalized link id 787 that is
public on the front page of neuroglancer.

I do not recall what the purpose of link id 791 was, but it is no longer
relevant as all cell annotations and the entire TG polygons are already
compiled in link id 787: https://www.brainsharer.org/ng/?id=787

As for changing the cell color, I was able to update all cell colors at the
same time on link id 791. I can show you how to change the color for both
polygon and cell annotations.

Best,
Marissa

**Comment by eddyod** [comment]
Litao has implemented a fix to change all points in a volume at one go.
